### PR TITLE
Fixes crash when undoing a multi-edit change in the DPE based Entity Inspector.

### PR DESCRIPTION
**Pull request by srikappa-amzn**
## What does this PR do?

This PR fixes the crash reported in https://github.com/o3de/o3de/issues/15918

Changes include:

- Listening to OnEntityDestruction() instead of OnEntityDestroyed() because the former gets sent when the entity destruction begins and the later is sent when it ends. Between the begin and the end, InvalidatePropertyDisplay() is getting called, which is queuing up a property refresh. So I changed it to listen to OnEntityDestruction
- Disconnected from buses when the entity is being destroyed. This prevents any further refreshes from being queued.
- Changed assert to if check for empty entity to handle cases where entity id is valid(eg. 0, 1, 2, etc.) but doesn't correspond to a live entity. Could happen when the component adapter holds garbage data.

## How was this PR tested?

Manually. Also verified that a previously fixed crash remains fixed : https://github.com/o3de/o3de/issues/14691

## Next steps

While the crash is fixed, the undo nodes are still not grouped. i.e, if you select 4 entities and make a multi-edit change and hit undo, you'll need to hit undo 4 times in order for that to work. Will look into fixing that next.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DPEComponentAdapter.h [hunk 1]
```diff
@@ -57,7 +57,9 @@ namespace AZ::DocumentPropertyEditor
         //! @param serializedPath The serialized path to use to check whether an override is present corresponding to it.
         void CreateLabel(AdapterBuilder* adapterBuilder, AZStd::string_view labelText, AZStd::string_view serializedPath) override;
 
-        void OnEntityDestroyed(const AZ::EntityId&) override;
+
```
Comment (forkercat) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DPEComponentAdapter.h:60]: nit: extra empty line

**Review by forkercat — APPROVED** [review]
Nice catch. Thanks for fixing this!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DPEComponentAdapter.cpp [hunk 2]
```diff
@@ -168,10 +173,14 @@ namespace AZ::DocumentPropertyEditor
...
 
-    void ComponentAdapter::OnEntityDestroyed(const AZ::EntityId& entityId)
+    void ComponentAdapter::OnEntityDestruction(const AZ::EntityId& entityId)
     {
         if (entityId == m_entityId)
         {
+            AzToolsFramework::PropertyEditorGUIMessages::Bus::Handler::BusDisconnect();
+            AzToolsFramework::ToolsApplicationEvents::Bus::Handler::BusDisconnect();
+            AzToolsFramework::PropertyEditorEntityChangeNotificationBus::MultiHandler::BusDisconnect(m_entityId);
+
```
Comment (michabr) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DPEComponentAdapter.cpp:183]: I'm wondering if some of these buses are expected to get called on entity destruction (ex. to update the outliner).

**Review by michabr — APPROVED** [review]
Thanks!

**Review by dtamkin1 — APPROVED** [review]

**Review by alexmontAmazon — APPROVED** [review]

**Pull request merged**